sdk/python/symbolchain/nem/MessageEncoder.py

```diff
@@ -0,0 +1,96 @@
+import secrets
+
+import cryptography
+
+from symbolchain.Cipher import AesCbcCipher, AesGcmCipher
+from symbolchain.CryptoTypes import PublicKey
+from symbolchain.nc import Message, MessageType
+from symbolchain.nem.KeyPair import KeyPair
+from symbolchain.nem.SharedKey import SharedKey
+
+
+class MessageEncoder:
+	"""Encrypts and encodes messages between two parties."""
+
+	def __init__(self, key_pair: KeyPair):
+		"""Creates message encoder around key pair."""
+		self.key_pair = key_pair
+
+	def _decode_aes_gcm(self, recipient_public_key, encoded_message):
+		# pylint: disable=duplicate-code
+		GCM_IV_SIZE = 12  # pylint: disable=invalid-name
+
+		tag = encoded_message[:AesGcmCipher.TAG_SIZE]
+		initialization_vector = encoded_message[AesGcmCipher.TAG_SIZE:AesGcmCipher.TAG_SIZE + GCM_IV_SIZE]
+		encoded_message_data = encoded_message[AesGcmCipher.TAG_SIZE + GCM_IV_SIZE:]
+
+		shared_key = SharedKey.derive_shared_key(self.key_pair, recipient_public_key)
+		cipher = AesGcmCipher(shared_key)
+		return cipher.decrypt(encoded_message_data + tag, initialization_vector)
+
+	def _decode_aes_cbc(self, recipient_public_key, encoded_message):
+		CBC_IV_SIZE = 16  # pylint: disable=invalid-name
+
+		salt = encoded_message[:32]
+		initialization_vector = encoded_message[32:32 + CBC_IV_SIZE]
+		encoded_message_data = encoded_message[32 + CBC_IV_SIZE:]
+		shared_key = SharedKey.derive_shared_key_deprecated(self.key_pair, recipient_public_key, salt)
+		cipher = AesCbcCipher(shared_key)
+		return cipher.decrypt(encoded_message_data, initialization_vector)
+
+	def try_decode(self, recipient_public_key, encoded_message: Message):
+		"""Tries to decode encoded message, returns tuple:
+		* True, message - if message has been decoded and decrypted
+		* False, encoded_message - otherwise
+		"""
+
+		if MessageType.ENCRYPTED != encoded_message.message_type:
+			raise RuntimeError('invalid message format')
```

Comment:
should this not accept PLAIN too ?

Reply:
not sure / have mixed feelings...  what would be return value(s) in case of symbol?

Reply:
if not, then user would always have to check before calling. not sure if that is best usability?

cc: @cryptoBeliever

Reply:
I'm not sure. If MessageEncoder is supporting just encrypted messages I guess we should raise an error or return `false, encoded_message`.